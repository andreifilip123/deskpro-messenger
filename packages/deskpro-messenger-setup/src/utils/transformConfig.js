export default (settings) => {
  const screens = {
    index: {
      screenType: 'Blocks',
      blocks: []
    }
  };

  if (settings.getIn(['chat', 'enabled'])) {
    const chatBlockConfig = settings.getIn(['messenger', 'chat']);
    screens.index.blocks.push({
      blockType: 'StartChatBlock',
      title: chatBlockConfig.get('title', 'Conversation'),
      description: chatBlockConfig.get('description'),
      linkText: chatBlockConfig.get('buttonText', 'Start a chat'),
      showAgentPhotos: chatBlockConfig.get('showAgentPhotos'),
      to: 'startChat'
    });
    screens.startChat = settings.get('chat').toJS();
    screens.startChat.screenType = 'StartChatScreen';
    delete screens.startChat.enabled;
    if (screens.startChat.noAnswerBehavior === 'save_ticket') {
      screens.startChat.ticketFormConfig = [
        {
          fields: [
            {
              name: 'name',
              label: 'Full Name',
              type: 'text',
              validation: ['required'],
              placeholder: 'John Doe'
            },
            {
              name: 'email',
              label: 'Email',
              type: 'text',
              validation: ['required'],
              placeholder: 'john.doe@company.com'
            }
          ]
        }
      ];
    }
  }

  if (settings.getIn(['tickets', 'enabled'])) {
    const ticketBlockConfig = settings.getIn(['messenger', 'tickets']);
    screens.index.blocks.push({
      blockType: 'ScreenLink',
      blockTitle: ticketBlockConfig.get('title'),
      description: ticketBlockConfig.get('description'),
      label: ticketBlockConfig.get('buttonText'),
      to: 'newTicket'
    });
    screens.newTicket = settings.get('tickets').toJS();
    delete screens.newTicket.enabled;
  }

  const enabledGreetings = [null];
  if ('object' === typeof screens.startChat) {
    enabledGreetings.push('/screens/startChat');
  }
  const config = {
    screens,
    themeVars: {
      '--color-primary': settings.getIn(['styles', 'primaryColor'], '#3d88f3'),
      '--color-background': settings.getIn(['styles', 'backgroundColor'], '#3d88f3'),
      '--brand-primary': settings.getIn(['styles', 'primaryColor'], '#3d88f3'),
      '--brand-secondary': settings.getIn(['styles', 'backgroundColor'], '#a9b0b0'),
      '--text-color': settings.getIn(['styles', 'textColor'], '#ffffff'),

    },
    greetings: {},
    enabledGreetings
  };

  return config;
};
