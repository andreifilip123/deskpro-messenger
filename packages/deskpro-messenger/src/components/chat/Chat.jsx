import React, { Fragment, PureComponent } from 'react';
import PropTypes from 'prop-types';
import { injectIntl } from 'react-intl';
import ScrollArea from 'react-scrollbar/dist/no-css';

import MessageBubble from './MessageBubble';
import SystemMessage from './SystemMessage';
import MessageForm from './MessageForm';
import TypingMessage from './TypingMessage';
import TranscriptBlock from './TranscriptBlock';
import RatingBlock from './RatingBlock';
import ChatEndBlock from './ChatEndBlock';
import SaveTicketBlock from './SaveTicketBlock';
import CreateTicketBlock from './CreateTicketBlock';
import { withScreenContentSize } from '../core/ScreenContent';
import { withFrameContext } from '../core/Frame';
import BotBubble from './BotBubble';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { isMessageFormFocused } from '../../modules/app';
import isMobile from 'is-mobile';
import { endBlockShown } from '../../modules/chat';

const mobile = isMobile();

const transMessages = {
  agentAssigned: {
    id: 'chat.agent_assigned.message',
    defaultMessage: `{name} joined the conversation.`
  },
  userLeft: {
    id: 'message_user-left',
    defaultMessage: `{name} left the conversation.`
  },
  noAgentOnline: {
    id: 'chat.no_agent_online',
    defaultMessage: 'Sorry, no one is online to accept your chat.'
  },
  chatEnded: {
    id: 'chat.ended',
    defaultMessage: 'Chat ended'
  }
};

const createTrans = ({ message, ...data }, type) => {
  if (typeof message === 'object' && message.phrase_id) {
    return [{ id: message.phrase_id }, message];
  } else if (type in transMessages) {
    return [transMessages[type], data];
  }
  return [];
};

class Chat extends PureComponent {
  static propTypes = {
    messages: PropTypes.array,
    onSendMessage: PropTypes.func.isRequired,
    onEndChat: PropTypes.func.isRequired,
    onCancelEndChat: PropTypes.func.isRequired,
    typing: PropTypes.oneOfType([PropTypes.object, PropTypes.bool]),
    agent: PropTypes.object,
    user: PropTypes.object,
    endChatBlock: PropTypes.bool.isRequired,
    chat: PropTypes.shape({
      id: PropTypes.number.isRequired,
      status: PropTypes.string.isRequired
    }),
    contentSize: PropTypes.shape({
      animating: PropTypes.bool,
      height: PropTypes.number,
      maxHeight: PropTypes.number
    }),
  };

  static defaultProps = {
    messages: [],
    agent: {},
    user: {},
    contentSize: {animating: false, height: undefined, maxHeight: undefined}
  };

  scrollArea = React.createRef();

  componentDidMount() {
    if (this.props.messages.length > 0 && this.scrollArea.current) {
      this.scrollToBottom();
    }
  }

  componentDidUpdate(prevProps) {
    const changedSize =
            prevProps.contentSize.height < this.props.contentSize.height;
    const lastPrevChat = prevProps.messages[prevProps.messages.length - 1];
    const lastCurrentChat = this.props.messages[this.props.messages.length - 1];

    if (
      !this.props.contentSize.animating && (
        prevProps.contentSize.animating ||
        prevProps.messages.length !== this.props.messages.length ||
        (lastPrevChat && lastPrevChat.id !== lastCurrentChat.id) ||
        changedSize || prevProps.endChatBlock !== this.props.endChatBlock
      )
    ) {
      this.scrollToBottom();
    }
  }

  scrollToBottom() {

    setTimeout(() => {
      if (this.scrollArea.current) {
        this.scrollArea.current.setState(
          {containerHeight: this.scrollArea.current.wrapper.offsetHeight},
          () => this.scrollArea.current.scrollBottom()
        );
      }
    }, 10);
  }

  render() {
    const {
      typing,
      messages,
      onSendMessage,
      onEndChat,
      onCancelEndChat,
      intl,
      agent,
      user,
      chat,
      chatConfig,
      endChatBlock,
      contentSize: { height, maxHeight },
      frameContext,
      formFocused,
      history
    } = this.props;

    const maxHeightAltered = maxHeight - (formFocused && mobile ? 207 : 240);

    return (
      <div
        className="dpmsg-ChatMessagesWrapper"
        style={{
          display: 'flex',
          flexDirection: 'column',
          height: maxHeightAltered
        }}
      >
        <ScrollArea
          horizontal={false}
          style={{
            flexBasis: 0,
            flexGrow: 1,
            overflow: height >= maxHeightAltered ? 'hidden' : undefined,
            minHeight: height >= maxHeightAltered ? '1px' : undefined
          }}
          ref={this.scrollArea}
          stopScrollPropagation={true}
          contentWindow={frameContext.window}
          ownerDocument={frameContext.document}
        >
          {!!chatConfig.prompt && (
            <BotBubble message={chatConfig.prompt} />
          )}
          {messages
            .filter(m => m.id || m.type !== 'chat.typing.start')
            .map((message, index) => {
              const key = message.uuid || `${message.type}-${index}`;
              switch (message.type) {
                case 'chat.message':
                  return <MessageBubble key={key} {...message} prev={messages[index-1]}/>;
                case 'chat.agentAssigned':
                case 'chat.agentUnassigned':
                  return null;
                case 'chat.ended':
                case 'chat.userTimeout':
                case 'chat.waitTimeout':
                  return chat.assigned && (
                    <Fragment key={key}>
                      <SystemMessage
                        {...message}
                        message={intl.formatMessage(
                          ...createTrans(message, 'chatEnded')
                        )}
                      />
                      <TranscriptBlock onSend={onSendMessage} user={user} />
                      <RatingBlock
                        onSend={onSendMessage}
                        user={user}
                        agent={agent}
                      />
                    </Fragment>
                  );
                case 'chat.noAgents':
                  return (
                    <div key={`no_agents_${message.uuid}`}>
                      <BotBubble
                        message={
                          chatConfig.busyMessage ||
                          intl.formatMessage(
                            ...createTrans(message, 'noAgentOnline')
                          )
                        }
                      />
                      {chatConfig.noAnswerBehavior === 'save_ticket' && (
                        <SaveTicketBlock
                          user={user}
                          uploadTo={chatConfig.uploadTo}
                          ticketParams={chatConfig.ticketDefaults}
                          formConfig={chatConfig.ticketFormConfig}
                          onSend={onSendMessage}
                          endChat={onEndChat}
                          history={history}
                        />
                      )}
                      {chatConfig.noAnswerBehavior === 'create_ticket' && (
                        <CreateTicketBlock chat={chat}/>
                      )}
                    </div>
                  );
                case 'chat.userJoined':
                case 'chat.userLeft':
                  return (
                    <SystemMessage
                      key={key}
                      {...message}
                      message={intl.formatMessage(...createTrans(message))}
                    />
                  );
                default: {
                  if (
                    message.origin === 'system' &&
                    message.message &&
                    message.message.phrase_id
                  ) {
                    return (
                      <BotBubble
                        key={key}
                        message={intl.formatMessage(...createTrans(message))}
                      />
                    );
                  }
                  return null;
                }
              }
          })}
          {!!typing && <TypingMessage value={typing} />}
          {!!chat && chat.status !== 'ended' && endChatBlock && (
            <ChatEndBlock onCancelEnd={onCancelEndChat} onEnd={onEndChat} />
          )}
        </ScrollArea>
        {!!chat && chat.status !== 'ended' && (
          <MessageForm
            frameContext={frameContext}
            onSend={onSendMessage}
            style={{ flex: '0 0 auto' }}
          />
        )}
      </div>
    );
  }
}

export default compose(
  injectIntl,
  withFrameContext,
  withScreenContentSize,
  connect(
    (state) => ({ formFocused: isMessageFormFocused(state), endChatBlock: endBlockShown(state) })
  )
)(Chat);



