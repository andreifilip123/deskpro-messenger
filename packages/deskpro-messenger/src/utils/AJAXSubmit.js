// eslint-disable-next-line func-names
const AJAXSubmit = (function () {
  function submitData(oData, config) {
    /* the AJAX request... */
    const oAjaxReq = new XMLHttpRequest();
    oAjaxReq.submittedData = oData;

    if (config.updateProgress) {
      oAjaxReq.addEventListener('progress', config.updateProgress);
    }
    if (config.transferComplete) {
      oAjaxReq.addEventListener('load', config.transferComplete);
    }
    if (config.transferFailed) {
      oAjaxReq.addEventListener('error', config.transferFailed);
    }
    if (config.transferCanceled) {
      oAjaxReq.addEventListener('abort', config.transferCanceled);
    }

    oAjaxReq.onreadystatechange = function() { // Call a function when the state changes.
      if (this.readyState === XMLHttpRequest.DONE) {
        if (this.status >= 400) {
          let message = this.statusText;
          if (this.response && this.response.errors && this.response.errors.file) {
            message = this.response.errors.file;
          }
          config.transferFailed({
            code: this.status,
            message
          });
        }
      }
    }

    oAjaxReq.withCredentials = true;
    /* method is POST */
    oAjaxReq.responseType = 'json';
    oAjaxReq.open('post', oData.receiver, true);
    /* enctype is multipart/form-data */
    const sBoundary = `---------------------------${Date.now().toString(16)}`;
    oAjaxReq.setRequestHeader('Content-Type', `multipart/form-data; boundary=${sBoundary}`);
    if (config.requestHeaders) {
      for (const [name, value] of Object.entries(config.requestHeaders)) {
        oAjaxReq.setRequestHeader(name, value);
      }
    }
    oAjaxReq.send(`--${sBoundary}\r\n${
      oData.segments.join(`--${sBoundary}\r\n`)}--${sBoundary}--\r\n`);
  }

  function processStatus(oData, config = {}) {
    if (oData.status > 0) { return; }
    /* the form is now totally serialized! do something before sending it to the server... */
    // /* doSomething(oData); */
    // /* console.log("AJAXSubmit - The form is now serialized. Submitting..."); */
    submitData(oData, config);
  }

  function pushSegment(oFREvt) {
    this.owner.segments[this.segmentIdx] += `${oFREvt.target.result}\r\n`;
    this.owner.status = this.owner.status - 1;
    processStatus(this.owner, this.config);
  }

  function SubmitRequest(config) {
    this.receiver = config.url;
    this.status = 0;
    this.segments = [];
    for (let nFile = 0; nFile < config.files.length; nFile++) {
      const oFile = config.files[nFile];
      const oSegmReq = new FileReader();
      /* (custom properties:) */
      oSegmReq.segmentIdx = this.segments.length;
      oSegmReq.owner = this;
      oSegmReq.config = config;
      /* (end of custom properties) */
      oSegmReq.onload = pushSegment;
      this.segments.push(`Content-Disposition: form-data; name="file"; filename="${oFile.name}"\r\nContent-Type: ${oFile.type}\r\n\r\n`);
      this.status = this.status + 1;
      oSegmReq.readAsBinaryString(oFile);
    }
    this.segments.push(`Content-Disposition: form-data;`);
    processStatus(this, config);
  }

  // eslint-disable-next-line func-names
  return function (config) {
    if (!config.url) { return; }
    new SubmitRequest(config);
  };
}());

export default AJAXSubmit;
