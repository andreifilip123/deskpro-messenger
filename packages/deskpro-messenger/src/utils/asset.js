window.parent.DESKPRO_MESSENGER_OPTIONS =
  window.parent.DESKPRO_MESSENGER_OPTIONS || {};

export default (path) => {
  let edited = '';
  if (window.parent.DESKPRO_MESSENGER_OPTIONS.baseUrl) {
    edited = window.parent.DESKPRO_MESSENGER_OPTIONS.baseUrl.replace(/\/$/, "");
  }
  return `${edited}/assets/${path}`;
};
