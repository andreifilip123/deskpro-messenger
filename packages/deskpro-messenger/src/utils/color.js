/**
 * Returns true if opposite color is light or not.
 * @param {string} color Color hash.
 * @link https://stackoverflow.com/a/35970186
 */
export const isLightColor = (color) => {
  let hex = color.startsWith('#') ? color.slice(1) : color;
  // convert 3-digit hex to 6-digits.
  if (hex.length === 3) {
    hex = hex[0] + hex[0] + hex[1] + hex[1] + hex[2] + hex[2];
  }
  if (hex.length !== 6) {
    throw new Error('Invalid HEX color.');
  }
  const r = parseInt(hex.slice(0, 2), 16),
    g = parseInt(hex.slice(2, 4), 16),
    b = parseInt(hex.slice(4, 6), 16);
  // http://stackoverflow.com/a/3943023/112731
  return r * 0.299 + g * 0.587 + b * 0.114 > 186;
};
