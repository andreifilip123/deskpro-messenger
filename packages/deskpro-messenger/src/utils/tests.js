import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { MemoryRouter } from 'react-router-dom';

const customRender = (node, { translations = {}, ...options } = {}) => {
  return render(
    <MemoryRouter>
      <IntlProvider locale="en-US" messages={translations}>
        {node}
      </IntlProvider>
    </MemoryRouter>,
    options
  );
};

// re-export everything
export * from 'react-testing-library';

// override render method
export { customRender as render };
