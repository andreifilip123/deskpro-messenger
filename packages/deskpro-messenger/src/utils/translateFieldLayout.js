/**
 * Translates all strings (labels) in a field layout config.
 *
 * @param {Array} config Field Layout config
 * @param {function} formatMessage Callback from react-intl to translate a string.
 *
 * @returns new translated field layout config.
 */
export default (config, formatMessage) => {
  const t = (str) => formatMessage({ id: str, defaultMessage: str });

  return config.map((layout) => ({
    ...layout,
    fields: layout.fields.map((field) => {
      const newField = { ...field };
      newField.label = t(field.label);
      if (newField.placeholder) {
        newField.placeholder = t(field.placeholder);
      }
      if (Array.isArray(newField.validation)) {
        newField.validation = field.validation.map(
          (rule) =>
            typeof rule === 'object'
              ? { ...rule, message: t(rule.message) }
              : rule
        );
      }
      if (Array.isArray(newField.options)) {
        newField.options = field.options.map((option) => ({
          ...option,
          label: t(option.label)
        }));
      }
      if (
        typeof newField.dataSource === 'object' &&
        Array.isArray(newField.dataSource.getOptions)
      ) {
        newField.dataSource.getOptions = field.dataSource.getOptions.map(
          (option) => ({
            ...option,
            label: t(option.label)
          })
        );
      }
      return newField;
    })
  }));
};
