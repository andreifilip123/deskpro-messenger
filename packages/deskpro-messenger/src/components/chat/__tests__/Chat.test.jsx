import React from 'react';
import { render } from '../../../utils/tests';

import Chat from '../Chat';

const messages = [
  {
    uuid: '1a2b',
    origin: 'user',
    type: 'chat.message',
    message: 'Hello World!',
    name: 'John Doe'
  },
  {
    uuid: '8cd76a',
    type: 'chat.noAgents',
    origin: 'system'
  },
  {
    uuid: '2c45',
    origin: 'system',
    type: 'chat.agentAssigned',
    message: { phrase_id: 'message_agent_join', name: 'Jon Snow' }
  },
  {
    uuid: '34f4',
    origin: 'agent',
    type: 'chat.message',
    message: 'How can I help you today?',
    name: 'Jon Snow'
  },
  {
    uuid: '545a',
    origin: 'user',
    type: 'chat.message',
    message: 'Another message from the user',
    name: 'John Doe'
  },
  {
    uuid: '1321',
    origin: 'system',
    type: 'chat.system_message',
    message: { phrase_id: 'system message from history without translation' }
  },
  {
    uuid: 'abcde',
    origin: 'system',
    type: 'chat.ended',
    message: { phrase_id: 'message_chat_end' }
  }
];

const translations = {
  message_agent_join: 'Agent {name} joined the chat',
  message_chat_end: 'Chat ended',
  chat_prompt: 'Thank you for start a chat!'
};

const onSendMessage = jest.fn();

const agent = { name: 'Jon Snow' };
const user = { name: 'John Doe' };
const chat = { id: 21, status: 'ended' };
const chatConfig = { prompt: 'chat_prompt', noAnswerBehavior: 'create_ticket' };
const typing = {
  name: 'Jon Snow',
  avatar: 'https://deskpro.com/images/jon-snow.png'
};

describe('<Chat />', () => {
  it('should render correctly', () => {
    const { asFragment } = render(
      <Chat {...{ messages, onSendMessage, agent, user, chat, chatConfig }} />,
      { translations }
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should display messages', () => {
    const { getByText, getByAltText } = render(
      <Chat
        {...{ messages, onSendMessage, agent, user, chat, chatConfig, typing }}
      />,
      { translations }
    );
    expect(getByText('Thank you for start a chat!')).toBeInTheDocument();
    expect(getByText('Hello World!')).toBeInTheDocument();
    expect(
      getByText('Sorry, no one is online to accept your chat.')
    ).toBeInTheDocument();
    expect(
      getByText(
        'You can submit a ticket instead and one of our agents will reply to you via email as soon as possible.'
      )
    ).toBeInTheDocument();
    const userAvatar = getByAltText('John Doe');
    expect(userAvatar).toBeInTheDocument();
    expect(userAvatar.nodeName).toBe('IMG');
    const agentAvatar = getByAltText('Jon Snow');
    expect(agentAvatar).toBeInTheDocument();
    expect(agentAvatar.nodeName).toBe('IMG');
    expect(getByText('How can I help you today?')).toBeInTheDocument();
    expect(getByText('Another message from the user')).toBeInTheDocument();
    expect(getByText('Agent Jon Snow joined the chat')).toBeInTheDocument();
    expect(
      getByText('system message from history without translation')
    ).toBeInTheDocument();
    expect(getByText('Chat ended')).toBeInTheDocument();
    expect(getByText('...')).toBeInTheDocument();
  });
});
