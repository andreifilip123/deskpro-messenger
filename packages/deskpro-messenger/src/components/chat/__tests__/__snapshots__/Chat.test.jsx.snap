// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chat /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="dpmsg-MessageBubbleRow is-incoming"
  >
    <div
      class="dpmsg-BubbleCol"
    >
      <div
        class="dpmsg-BubbleItem"
      >
        Thank you for start a chat!
      </div>
    </div>
  </div>
  <div
    class="dpmsg-MessageBubbleRow is-outgoing"
  >
    <div
      class="dpmsg-AvatarCol is-rounded"
    >
      <img
        alt="John Doe"
        src="/assets/img/docs/avatar-default.jpg"
      />
    </div>
    <div
      class="dpmsg-BubbleCol"
    >
      <div
        class="dpmsg-BubbleItem"
      >
        Hello World!
      </div>
    </div>
  </div>
  <span
    class="dpmsg-BubbleNotification is-joined"
  >
    Sorry, no one is online to accept your chat.
  </span>
  <div
    class="dpmsg-MessagePrompt"
  >
    <div
      class="dpmsg-PromptHeader"
    >
      <span
        class="dpmsg-PromptHeaderText"
      >
        Submit a ticket
      </span>
    </div>
    <div
      class="dpmsg-PromptText"
    >
      <span>
        You can submit a ticket instead and one of our agents will reply to you via email as soon as possible.
      </span>
      <a
        class="dpmsg-Button Button--primary"
        href="/screens/newTicket"
      >
        <span>
          Create Ticket
        </span>
      </a>
    </div>
  </div>
  <span
    class="dpmsg-BubbleNotification is-joined"
  >
    Agent Jon Snow joined the chat
  </span>
  <div
    class="dpmsg-MessageBubbleRow is-incoming"
  >
    <div
      class="dpmsg-AvatarCol"
    >
      <img
        alt="Jon Snow"
        src="/assets/img/docs/avatar-default.jpg"
      />
    </div>
    <div
      class="dpmsg-BubbleCol"
    >
      <div
        class="dpmsg-BubbleItem"
      >
        How can I help you today?
      </div>
    </div>
  </div>
  <div
    class="dpmsg-MessageBubbleRow is-outgoing"
  >
    <div
      class="dpmsg-AvatarCol is-rounded"
    >
      <img
        alt="John Doe"
        src="/assets/img/docs/avatar-default.jpg"
      />
    </div>
    <div
      class="dpmsg-BubbleCol"
    >
      <div
        class="dpmsg-BubbleItem"
      >
        Another message from the user
      </div>
    </div>
  </div>
  <span
    class="dpmsg-BubbleNotification is-joined"
  >
    system message from history without translation
  </span>
  <span
    class="dpmsg-BubbleNotification is-joined"
  >
    Chat ended
  </span>
  <div
    class="dpmsg-MessagePrompt"
  >
    <div
      class="dpmsg-PromptHeader"
    >
      <i
        class="dpmsg-IconNotes"
      />
      <span
        class="dpmsg-PromptHeaderText"
      >
        Would you like an email transcript?
      </span>
    </div>
    <div
      class="dpmsg-PromptContentAgree"
    >
      <button
        class="dpmsg-PromptBtn is-agdee"
        name="yes"
      >
        <span>
          Yes
        </span>
      </button>
      <button
        class="dpmsg-PromptBtn is-disagree"
        name="no"
      >
        <span>
          No
        </span>
      </button>
    </div>
  </div>
  <div
    class="dpmsg-MessagePrompt"
  >
    <div
      class="dpmsg-PromptHeader"
    >
      <span
        class="dpmsg-PromptHeaderText"
      >
        Rate your conversation with Jon Snow
      </span>
    </div>
    <div
      class="dpmsg-PromptContentEvaluation"
    >
      <button
        class="dpmsg-Button Button-LimitedWidth Button--success"
        name="helpful"
      >
        Helpful
      </button>
      <button
        class="dpmsg-Button Button-LimitedWidth Button--danger"
        name="unhelpful"
      >
        Unhelpful
      </button>
    </div>
  </div>
</DocumentFragment>
`;
